src/transports/ssh_libssh2.c

```diff
@@ -0,0 +1,437 @@
+	return -1;
+}
+
+git_ssh_channel *git_ssh_channel_open(git_ssh_session *s)
+{
+	git_ssh_channel *channel;
+
+	channel = git__calloc(1, sizeof(*channel));
+	if (!channel) {
+		git_error_set_oom();
+		return NULL;
+	}
+
+	channel->channel = libssh2_channel_open_session(s->session);
+
+	if (channel->channel)
+		libssh2_channel_set_blocking(channel->channel, 1);
+
+	return channel;
+}
+
+void git_ssh_channel_free(git_ssh_channel *c)
+{
+	libssh2_channel_free(c->channel);
+	git__free(c);
+}
+
+int git_ssh_channel_read(char *buffer, size_t size, int is_stderr, git_ssh_channel *c)
+{
+	if (is_stderr)
+		return libssh2_channel_read_stderr(c->channel, buffer, size);
+	return libssh2_channel_read(c->channel, buffer, size);
+}
+
+int git_ssh_channel_write(git_ssh_channel *c, const char *buffer, size_t size)
+{
+	return libssh2_channel_write(c->channel, buffer, size);
+}
+
+int git_ssh_channel_exec(git_ssh_channel *c, const char *request)
+{
+	return libssh2_channel_exec(c->channel, request);
+}
+
+int git_ssh_session_server_hostkey(git_ssh_session *s, git_cert_hostkey *cert)
+{
+	const char *key;
+	size_t cert_len;
+	int cert_type;
+
+	cert->parent.cert_type = GIT_CERT_HOSTKEY_LIBSSH2;
+
+	key = libssh2_session_hostkey(s->session, &cert_len, &cert_type);
+	if (key != NULL) {
+		cert->type |= GIT_CERT_SSH_RAW;
+		cert->hostkey = key;
+		cert->hostkey_len = cert_len;
+		switch (cert_type) {
+			case LIBSSH2_HOSTKEY_TYPE_RSA:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_RSA;
+				break;
+			case LIBSSH2_HOSTKEY_TYPE_DSS:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_DSS;
+				break;
+
+#ifdef LIBSSH2_HOSTKEY_TYPE_ECDSA_256
+			case LIBSSH2_HOSTKEY_TYPE_ECDSA_256:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_KEY_ECDSA_256;
+				break;
+			case LIBSSH2_HOSTKEY_TYPE_ECDSA_384:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_KEY_ECDSA_384;
+				break;
+			case LIBSSH2_KNOWNHOST_KEY_ECDSA_521:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_KEY_ECDSA_521;
+				break;
+#endif
+
+#ifdef LIBSSH2_HOSTKEY_TYPE_ED25519
+			case LIBSSH2_HOSTKEY_TYPE_ED25519:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_KEY_ED25519;
+				break;
+#endif
+			default:
+				cert->raw_type = GIT_CERT_SSH_RAW_TYPE_UNKNOWN;
+		}
+	}
+
+#ifdef LIBSSH2_HOSTKEY_HASH_SHA256
+	key = libssh2_hostkey_hash(s->session, LIBSSH2_HOSTKEY_HASH_SHA256);
+	if (key != NULL) {
+		cert->type |= GIT_CERT_SSH_SHA256;
+		memcpy(&cert->hash_sha256, key, 32);
+	}
+#endif
+
+
+	key = libssh2_hostkey_hash(s->session, LIBSSH2_HOSTKEY_HASH_SHA1);
+	if (key != NULL) {
+		cert->type |= GIT_CERT_SSH_SHA1;
+		memcpy(&cert->hash_sha1, key, 20);
+	}
+
+	key = libssh2_hostkey_hash(s->session, LIBSSH2_HOSTKEY_HASH_MD5);
+	if (key != NULL) {
+		cert->type |= GIT_CERT_SSH_MD5;
+		memcpy(&cert->hash_md5, key, 16);
+	}
+
+	if (cert->type == 0) {
+		git_error_set(GIT_ERROR_SSH, "unable to get the host key");
+		return -1;
+	}
+
+	return 0;
+}
+
+int git_ssh_session_server_is_known(git_ssh_session *s, int *valid)
+{
+	GIT_UNUSED(s);
+	*valid = 0;
```

Comment:
This ensures that libssh2's behavior doesn't change (default to "can't say it's valid", a.k.a `0`), but now libssh checks for known_hosts validity, while libssh2 doesn't (and it looks messy to handle appropriately). I'm not even sure if it makes sense to update the known_host file afterwards…